File: src/notification_tray/components/notification_widget.py
import logging

from PyQt5.QtCore import QEvent, Qt, QTimer, pyqtSignal
from PyQt5.QtGui import QIcon, QImage, QMouseEvent, QPixmap
from PyQt5.QtWidgets import (QHBoxLayout, QLabel, QPushButton, QVBoxLayout,
                             QWidget)

from ..types.notification import CachedNotification
from .notification_service import NotificationCloseReason

logger = logging.getLogger(__name__)

import os

default_timeout = 5000
try:
    default_timeout = (
        max(0, int(os.environ["NOTIFICATION_TRAY_DEFAULT_TIMEOUT_MILLIS"])) or 5000
    )
except:
    pass


class NotificationWidget(QWidget):
    action_invoked = pyqtSignal(str)
    displayed = pyqtSignal()
    closed = pyqtSignal(int)

    def __init__(self, data: CachedNotification):
        super().__init__()
        self.data = data
        self.was_displayed = False
        self.displayed.connect(self._set_was_displayed)
        self.initUI()
        self.timer = QTimer(self)
        self.timer_remaining_time = -1
        self.timer.setSingleShot(True)
        self.timer.timeout.connect(
            lambda: self.closed.emit(NotificationCloseReason.EXPIRED.value)
        )

    def mousePressEvent(self, a0: QMouseEvent | None) -> None:
        actions = self.data["actions"]
        default_action = (
            next(iter(actions))
            if len(actions) == 1
            else "default" if "default" in actions else None
        )
        if default_action:
            self.action_invoked.emit(default_action)
        return super().mousePressEvent(a0)

    def enterEvent(self, a0: QEvent | None) -> None:
        self.timer_remaining_time = self.timer.remainingTime()
        self.timer.stop()
        return super().enterEvent(a0)

    def leaveEvent(self, a0: QEvent | None) -> None:
        self.timer.start(self.timer_remaining_time)
        return super().leaveEvent(a0)

    def _set_was_displayed(self):
        self.was_displayed = True

    def _get_icon(self) -> QIcon | None:
        if image_data := (
            self.data["hints"].get("image-data")
            or self.data["hints"].get("image_data")
            or self.data["hints"].get("icon_data")
        ):
            logger.debug("Getting notification icon from hints")
            width, height, rowstride, has_alpha, bits_per_sample, channels, data = (
                image_data
            )
            assert bits_per_sample == 8
            assert channels == 4 if has_alpha else 3
            return QIcon(
                QPixmap.fromImage(
                    QImage(
                        bytes(data),
                        width,
                        height,
                        rowstride,
                        (
                            QImage.Format.Format_RGBA8888
                            if has_alpha
                            else QImage.Format.Format_RGB888
                        ),
                    )
                )
            )

        elif label := (
            self.data["hints"].get("image-path")
            or self.data["hints"].get("image_path")
            or self.data["app_icon"]
        ):
            if label.startswith("file://"):
                logger.debug(f"Getting notification icon from path {label}")
                return QIcon(label.removesuffix("file://"))
            else:
                logger.debug(
                    f"Getting notification icon from theme with icon name {label}"
                )
                return QIcon.fromTheme(label)

    def initUI(self):
        logger.info("Creating notification widget")
        self.setFixedWidth(256)
        self.setWindowFlags(
            Qt.WindowType.WindowStaysOnTopHint
            | Qt.WindowType.FramelessWindowHint
            | Qt.WindowType.ToolTip  # type: ignore
        )
        self.setAttribute(Qt.WidgetAttribute.WA_ShowWithoutActivating)

        layout = QVBoxLayout()
        top_layout = QHBoxLayout()

        logger.debug("Getting icon")
        try:
            icon = self._get_icon()
        except Exception as e:
            logger.error(f"Could not get icon: {e}")
            icon = None
        if icon:
            logger.debug("Got icon")
            icon_label = QLabel(self)
            icon_label.setAlignment(Qt.AlignmentFlag.AlignLeft)
            icon_label.setPixmap(icon.pixmap(32, 32))
            top_layout.addWidget(icon_label)
        else:
            logger.debug("No icon")

        app_name_label = QLabel(self.data["app_name"], self)
        app_name_label.setObjectName(
            "appLabel"
        )  # for compatibility with lxqt-notificationd themes
        top_layout.addWidget(app_name_label)

        close_button = QPushButton()
        close_button.setObjectName(
            "closeButton"
        )  # for compatibility with lxqt-notificationd themes
        close_button.clicked.connect(
            lambda: self.closed.emit(NotificationCloseReason.DISMISSED_BY_USER.value)
        )

        top_layout.addWidget(close_button)
        layout.addLayout(top_layout)
        summary_layout = QVBoxLayout()
        summary_label = QLabel(self.data["summary"] + "\n" + self.data["body"], self)
        summary_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        summary_label.setWordWrap(True)
        summary_layout.addWidget(summary_label)
        layout.addLayout(summary_layout)

        button_layout = QHBoxLayout()

        logger.debug("Adding action buttons")
        for key, value in self.data["actions"].items():
            button = QPushButton(None, self)
            if self.data["hints"].get("action-icons"):
                button.setIcon(QIcon.fromTheme(value))
            else:
                button.setText(value)
            button.clicked.connect(lambda: self.action_invoked.emit(key))
            logger.debug(f"Adding action button {key} => {value}")
            button_layout.addWidget(button)

        layout.addLayout(button_layout)
        self.setLayout(layout)
        self.adjustSize()

        if self.data["expire_timeout"] and self.data["hints"].get("urgency") != 2:
            self.displayed.connect(self.schedule_close)

    def schedule_close(self):
        timeout = (
            default_timeout
            if self.data["expire_timeout"] == -1
            else self.data["expire_timeout"]
        )
        logger.info(
            f"Scheduling close of notification {self.data['id']} in {timeout / 1000} seconds"
        )
        self.timer.start(timeout)

File: src/notification_widget.cpp
#include "notification_widget.h"

#include "ui_notification_widget.h"
#include "utils/logging.h"

#include <QDBusArgument>
#include <QFile>
#include <QHBoxLayout>
#include <QIcon>
#include <QImage>
#include <QLabel>
#include <QMenu>
#include <QMouseEvent>
#include <QPainter>
#include <QPaintEvent>
#include <QPixmap>
#include <QPushButton>
#include <QStyle>
#include <QStyleOption>
#include <QToolButton>
#include <QUrl>
#include <QVBoxLayout>

#include <cstdlib>

static Logger logger = Logger::getLogger("NotificationWidget");

#define ICONSIZE QSize(32, 32)

static int getDefaultTimeout() {
    const char* env = std::getenv("NOTIFICATION_TRAY_DEFAULT_TIMEOUT_MILLIS");
    if (env) {
        int timeout = std::atoi(env);
        return (timeout > 0) ? timeout : 5000;
    }
    return 5000;
}

NotificationWidget::NotificationWidget(const CachedNotification& data, QWidget* parent)
    : QWidget(parent)
    , data(data)
    , ui(new Ui::NotificationWidget)
    , m_timer(nullptr) {
    setupUi();
    setValues();

    connect(this, &NotificationWidget::displayed, this, &NotificationWidget::setWasDisplayed);
}

NotificationWidget::~NotificationWidget() {
    delete ui;
}

void NotificationWidget::setupUi() {
    ui->setupUi(this);

    // Set object name to "Notification" so LXQt theme stylesheets apply correctly
    // (stylesheets use "Notification" selector, not "NotificationWidget")
    setObjectName("Notification");

    setWindowFlags(Qt::WindowStaysOnTopHint | Qt::FramelessWindowHint | Qt::ToolTip);
    setAttribute(Qt::WA_ShowWithoutActivating);
    setMouseTracking(true);

    connect(ui->closeButton, &QToolButton::clicked, this, &NotificationWidget::closeButton_clicked);

    // Setup settings button with snooze menu
    QMenu* settingsMenu = new QMenu(this);
    QMenu* snoozeMenu = settingsMenu->addMenu("Snooze");

    QAction* snooze1min = snoozeMenu->addAction("1 minute");
    QAction* snooze5min = snoozeMenu->addAction("5 minutes");
    QAction* snooze30min = snoozeMenu->addAction("30 minutes");

    connect(snooze1min, &QAction::triggered, [this]() { snoozeNotification(60000); });
    connect(snooze5min, &QAction::triggered, [this]() { snoozeNotification(300000); });
    connect(snooze30min, &QAction::triggered, [this]() { snoozeNotification(1800000); });

    ui->settingsButton->setMenu(settingsMenu);
}

void NotificationWidget::setValues() {
    // Icon handling - following lxqt-notificationd's priority order:
    // 1. "image-data" hint
    // 2. "image_data" hint (compatibility)
    // 3. "image-path" hint
    // 4. "image_path" hint (compatibility)
    // 5. app_icon parameter
    // 6. "icon_data" hint (compatibility)

    QPixmap pixmap;

    if (!data.hints.value("image-data").isNull()) {
        pixmap = getPixmapFromHint(data.hints.value("image-data"));
    } else if (!data.hints.value("image_data").isNull()) {
        pixmap = getPixmapFromHint(data.hints.value("image_data"));
    } else if (!data.hints.value("image-path").isNull()) {
        pixmap = getPixmapFromString(data.hints.value("image-path").toString());
    } else if (!data.hints.value("image_path").isNull()) {
        pixmap = getPixmapFromString(data.hints.value("image_path").toString());
    } else if (!data.app_icon.isEmpty()) {
        pixmap = getPixmapFromString(data.app_icon);
    } else if (!data.hints.value("icon_data").isNull()) {
        pixmap = getPixmapFromHint(data.hints.value("icon_data"));
    }

    // Hide icon if not found (matches lxqt behavior - issue #325)
    if (pixmap.isNull()) {
        ui->iconLabel->hide();
    } else {
        if (pixmap.size().width() > ICONSIZE.width() ||
            pixmap.size().height() > ICONSIZE.height()) {
            pixmap = pixmap.scaled(ICONSIZE, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        }
        ui->iconLabel->setPixmap(pixmap);
        ui->iconLabel->show();
    }

    // Application name
    ui->appLabel->setVisible(!data.app_name.isEmpty());
    ui->appLabel->setText(data.app_name);

    // Summary - hide if same as app name
    ui->summaryLabel->setVisible(!data.summary.isEmpty() && data.app_name != data.summary);
    ui->summaryLabel->setText(data.summary);

    // Body - convert newlines to <br/> for proper HTML rendering
    ui->bodyLabel->setVisible(!data.body.isEmpty());
    QString formattedBody = data.body;
    formattedBody.replace('\n', "<br/>");
    ui->bodyLabel->setText(formattedBody);

    // Timer setup
    if (m_timer) {
        m_timer->stop();
        m_timer->deleteLater();
        m_timer = nullptr;
    }

    int urgency = data.hints.value("urgency", 1).toInt();
    // Don't auto-close critical notifications (urgency == 2) or those with timeout == 0
    if (data.expire_timeout != 0 && urgency != 2) {
        m_timer = new NotificationTimer(this);
        connect(m_timer, &NotificationTimer::timeout,
                [this]() { emit closed(static_cast<int>(NotificationCloseReason::EXPIRED)); });
        connect(this, &NotificationWidget::displayed, this, &NotificationWidget::scheduleClose);
    }

    // Actions - add buttons to actionsLayout
    // Match lxqt-notificationd's behavior: use NoFocus and set objectName
    for (const auto& [key, value] : data.actions) {
        QPushButton* button = new QPushButton(this);
        button->setObjectName(key);
        button->setFocusPolicy(Qt::NoFocus);
        if (data.hints.value("action-icons", false).toBool()) {
            button->setIcon(QIcon::fromTheme(value));
        } else {
            button->setText(value);
        }
        connect(button, &QPushButton::clicked, [this, key = key]() { emit actionInvoked(key); });
        ui->actionsLayout->addWidget(button);
    }

    adjustSize();
}

QPixmap NotificationWidget::getPixmapFromHint(const QVariant& argument) const {
    // Image data format from D-Bus spec:
    // (width, height, rowstride, has_alpha, bits_per_sample, channels, data)

    if (!argument.canConvert<QDBusArgument>()) {
        logger.warning("Image hint is not a QDBusArgument");
        return QPixmap();
    }

    int width, height, rowstride, bitsPerSample, channels;
    bool hasAlpha;
    QByteArray pixelData;

    const QDBusArgument arg = argument.value<QDBusArgument>();
    arg.beginStructure();
    arg >> width;
    arg >> height;
    arg >> rowstride;
    arg >> hasAlpha;
    arg >> bitsPerSample;
    arg >> channels;
    arg >> pixelData;
    arg.endStructure();

    bool rgb = !hasAlpha && channels == 3 && bitsPerSample == 8;
    QImage::Format imageFormat = rgb ? QImage::Format_RGB888 : QImage::Format_ARGB32;

    QImage img =
        QImage(reinterpret_cast<const uchar*>(pixelData.constData()), width, height, imageFormat);

    if (!rgb) {
        img = img.rgbSwapped();
    }

    return QPixmap::fromImage(img);
}

QPixmap NotificationWidget::getPixmapFromString(const QString& str) const {
    QUrl url(str);

    // Try as file URL first
    if (url.isValid() && QFile::exists(url.toLocalFile())) {
        logger.debug(QString("Loading icon from file URL: %1").arg(url.toLocalFile()));
        return QPixmap(url.toLocalFile());
    }

    // Try as file path (handles file:// prefix)
    if (str.startsWith("file://")) {
        QString path = str.mid(7);
        if (QFile::exists(path)) {
            logger.debug(QString("Loading icon from path: %1").arg(path));
            return QPixmap(path);
        }
    }

    // Try as absolute path
    if (QFile::exists(str)) {
        logger.debug(QString("Loading icon from path: %1").arg(str));
        return QPixmap(str);
    }

    // Fall back to theme icon
    logger.debug(QString("Loading icon from theme: %1").arg(str));
    QIcon themeIcon = QIcon::fromTheme(str);
    if (!themeIcon.isNull()) {
        return themeIcon.pixmap(ICONSIZE);
    }

    return QPixmap();
}

void NotificationWidget::mousePressEvent(QMouseEvent* event) {
    QString defaultAction;
    if (data.actions.size() == 1) {
        defaultAction = data.actions.begin()->first;
    } else if (data.actions.count("default")) {
        defaultAction = "default";
    }

    if (!defaultAction.isEmpty()) {
        emit actionInvoked(defaultAction);
    }

    QWidget::mousePressEvent(event);
}

void NotificationWidget::enterEvent(QEvent* event) {
    if (m_timer) {
        m_timer->pause();
    }
    QWidget::enterEvent(event);
}

void NotificationWidget::leaveEvent(QEvent* event) {
    if (m_timer) {
        m_timer->resume();
    }
    QWidget::leaveEvent(event);
}

void NotificationWidget::setWasDisplayed() {
    was_displayed = true;
}

void NotificationWidget::closeButton_clicked() {
    if (m_timer) {
        m_timer->stop();
    }
    emit closed(static_cast<int>(NotificationCloseReason::DISMISSED_BY_USER));
}

void NotificationWidget::snoozeNotification(int duration_ms) {
    logger.info(
        QString("Snoozing notification %1 for %2 seconds").arg(data.id).arg(duration_ms / 1000.0));
    if (m_timer) {
        m_timer->stop();
    }
    emit snoozed(duration_ms);
    hide();
}

void NotificationWidget::scheduleClose() {
    int timeout = (data.expire_timeout == -1) ? getDefaultTimeout() : data.expire_timeout;
    logger.info(QString("Scheduling close of notification %1 in %2 seconds")
                    .arg(data.id)
                    .arg(timeout / 1000.0));
    if (m_timer) {
        m_timer->start(timeout);
    }
}

void NotificationWidget::paintEvent(QPaintEvent*) {
    // Required for stylesheet backgrounds to work correctly on QWidget subclasses
    // (from lxqt-notificationd's Notification::paintEvent)
    QStyleOption opt;
    opt.initFrom(this);
    QPainter p(this);
    style()->drawPrimitive(QStyle::PE_Widget, &opt, &p, this);
}
